### ISSUE TITLE
Missing Entry in FIPS186-5 RSA KeyGen / "probableWithProvableAux" Test Vector Group Object

### ISSUE THREAD

**Issue by Snap-A** [usnistgov/ACVP#1514]
#### Test Request
Test Server:   Demo
Test Session: /acvp/v1/testSessions/512622
Test Group Id: 1

#### Description
The FIPS186-5 RSA key generation AFT test "probableWithProvableAux" is requiring the array fields `hashAlg` and `primeTest` to request test vectors on demo, as expected. (This matches the previous registration for FIPS186-4, test "B.3.5").
The IUT will then parse the group object for the two chosen entries, selected from each array in order to run the correct algorithm.

The parsing step fails, since the group object does _not_ contain the field `hashAlg`, just the field `primeTest`.
This seems to be a _probable bug_ in the server code, incorrectly omitting this entry.

#### JSON Samples
Registration
```
[                                                                                                                                   
    {                                                                                                                               
        "algorithm": "RSA",                                                                                                         
        "mode": "keyGen",                                                                                                           
        "revision": "FIPS186-5",                                                                                                    
        "pubExpMode": "fixed",                                                                                                      
        "fixedPubExp": "010001",                                                                                                    
        "infoGeneratedByServer": true,                                                                                              
        "keyFormat": "standard",                                                                                                    
        "capabilities": [                                                                                                           
            {                                                                                                                       
                "randPQ": "probableWithProvableAux",                                                                                
                "properties": [                                                                                                     
                    {                                                                                                               
                        "modulo": 2048,                                                                                             
                        "hashAlg": [                                                                                                
                            "SHA-1",                                                                                                
                            "SHA2-256"                                                                                              
                        ],                                                                                                          
                        "primeTest": [                                                                                              
                            "2pow100"                                                                                               
                        ]                                                                                                           
                    },                                                                                                              
                    {                                                                                                               
                        "modulo": 3072,                                                                                             
                        "hashAlg": [                                                                                                
                            "SHA-1",                                                                                                
                            "SHA2-256"                                                                                              
                        ],                                                                                                          
                        "primeTest": [                                                                                              
                            "2pow100"                                                                                               
                        ]                                                                                                           
                    },
                   {                                                                                                               
                        "modulo": 4096,                                                                                             
                        "hashAlg": [                                                                                                
                            "SHA-1",                                                                                                
                            "SHA2-256"                                                                                              
                        ],                                                                                                          
                        "primeTest": [                                                                                              
                            "2pow100"                                                                                               
                        ]                                                                                                           
                    }                                                                                                               
                ]                                                                                                                   
            }                                                                                                                       
        ]                                                                                                                           
    }                                                                                                                               
] 
```
Returned Test Vectors (First Group)
```
[ {
    "jwt": "<text>",
    "url": "\/acvp\/v1\/testSessions\/512622",
    "isSample": false,
    "vectorSetUrls": [
        "\/acvp\/v1\/testSessions\/512622\/vectorSets\/2348761"
    ]
}, {
    "vsId": 2348761,
    "algorithm": "RSA",
    "mode": "keyGen",
    "revision": "FIPS186-5",
    "isSample": false,
    "testGroups": [
        {
            "tgId": 1,
            "infoGeneratedByServer": true,
            "modulo": 2048,
            "testType": "AFT",
            "keyFormat": "standard",
            "primeTest": "2pow100",
            "randPQ": "probableWithProvableAux",
            "pubExp": "fixed",
            "fixedPubExp": "010001",
            "tests": [
                {
                    "tcId": 1,
                    "seed": "460064DA3A40E15AABD37E370F2754024103D0C6742F6A53904E6A49",
                    "bitlens": [
                        372,
                        304,
                        248,
                        400
                    ],
                    "xP": "<text>",
                    "xQ": "<text>",
                    "e": "010001"
                },
...
```

**Comment by jbrock24** [comment]
Hi @Snap-A , thanks for the feedback and help. We' have some time issues lately, but those are done, and I'll get to this by the middle of next week, if not sooner.

- Joel

**Comment by GlennUL** [comment]
Hi @jbrock24 , it looks like this didn't make it into today's release to Demo. Is there any eta for when this may be fixed?

**Comment by jbrock24** [comment]
Hi @GlennUL Unfortunately it didn't, the fix is in dev, but will go out with the next release, apologies.

**Comment by jbrock24** [comment]
Hi @Snap-A @GlennUL - After reviewing this, this is something that isn't required based on the sections 3.3 and 3.6. Within the code [here](https://github.com/usnistgov/ACVP-Server/blob/master/gen-val/src/generation/src/NIST.CVP.ACVTS.Libraries.Generation/RSA/Fips186_5/KeyGen/ContractResolvers/PromptProjectionContractResolver.cs#L79-L88) we specifically ignore this when outputting for this PrimeGenMode to the prompt file. 

```C#
// Ignore HashAlg for B.3.3, B.3.6
                if (testGroup.PrimeGenMode == PrimeGenModes.RandomProbablePrimes ||
                    testGroup.PrimeGenMode == PrimeGenModes.RandomProbablePrimesWithAuxiliaryProvablePrimes)
                {
                    if (jsonProperty.UnderlyingName.Equals(nameof(TestGroup.HashAlgName),
                        StringComparison.OrdinalIgnoreCase))
                    {
                        return false;
                    }
                }
```

Sorry for all the confusion regarding this. I'll leave it open for further conversation.

**Comment by jbrock24** [comment]
Feel free to close when ready.

**Comment by jbrock24** [comment]
Also check A.1.5 Generation of Probable Primes with Conditions Based on Auxiliary
Provable Primes section of [185-6](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-5.pdf).

**Comment by Snap-A** [comment]
> Hi @Snap-A @GlennUL - After reviewing this, this is something that isn't required based on the sections 3.3 and 3.6. Within the code [here](https://github.com/usnistgov/ACVP-Server/blob/master/gen-val/src/generation/src/NIST.CVP.ACVTS.Libraries.Generation/RSA/Fips186_5/KeyGen/ContractResolvers/PromptProjectionContractResolver.cs#L79-L88) we specifically ignore this when outputting for this PrimeGenMode to the prompt file.
> 
> ```cs
> // Ignore HashAlg for B.3.3, B.3.6
>                 if (testGroup.PrimeGenMode == PrimeGenModes.RandomProbablePrimes ||
>                     testGroup.PrimeGenMode == PrimeGenModes.RandomProbablePrimesWithAuxiliaryProvablePrimes)
>                 {
>                     if (jsonProperty.UnderlyingName.Equals(nameof(TestGroup.HashAlgName),
>                         StringComparison.OrdinalIgnoreCase))
>                     {
>                         return false;
>                     }
>                 }
> ```
> 
> Sorry for all the confusion regarding this. I'll leave it open for further conversation.

I believe this is in contradiction with the required _registration_ parameters. The attempt to register without the `hashAlg` value (array of hash names) fails.

This seems to point to an "interpretation" problem for this mode in the text you referenced. The mode under discussion creates "probable" primes, but *also* employs a "provable" auxiliary function. This auxiliary function uses a hash in its algorithm that needs to be set to successfully generate keys. Therefore, just using the the fact that "probable" primes are generated does _not_ preclude the test vector from providing hash identifiers.

Can you find out whether the needs of the auxiliary function were fully considered by the quoted reference?

**Comment by GlennUL** [comment]
Hi @jbrock24 Do you have any update on the vendors additional comments above?

**Comment by jbrock24** [comment]
Hi @Snap-A @GlennUL - Thanks for this guys, sorry for the confusion, but we got it worked through. I'll have this fix out in the *.35 hotfix.

**Comment by GlennUL** [comment]
Thanks Joel!

**Comment by Snap-A** [comment]
Yes, thank you Joel!

**Comment by livebe01** [comment]
The fix for this is on Demo in release [v1.1.0.35](https://github.com/usnistgov/ACVP-Server/releases/tag/v1.1.0.35).

**Comment by Snap-A** [comment]
The "hashAlgo" data is now included and the RSA test processing in my ITU is working.
Thank you!

**Comment by jbrock24** [comment]
Great, thank you for your patience! I'll close this out now.

**Comment by Snap-A** [comment]
Is there a reason this ticket was re-opened? 
Also, I presume everything changed on "demo" will be pushed to "prod" in the future.

**Comment by livebe01** [comment]
Sure, Joel inadvertently closed the ticket last week. We like to keep tickets open until the fix has been deployed across all environments (Demo and Prod). 

That's correct. When the [v1.1.0.35](https://github.com/usnistgov/ACVP-Server/releases/tag/v1.1.0.35) release gets deployed to Prod tomorrow, Demo and Prod will be in sync.

**Comment by jbrock24** [comment]
Still new to the process, sorry @Snap-A :D

**Comment by livebe01** [comment]
The fix for this is on Prod in release [v1.1.0.35](https://github.com/usnistgov/ACVP-Server/releases/tag/v1.1.0.35).
